### PR TITLE
Feature: Watch api

### PR DESCRIPTION
**Pull request by lichuang**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

1. add watch API for metasrv
2. add unit tests for watch API.

## Changelog

- New Feature

## Related Issues

Fixes #issue

### CODE REVIEW

File: common/meta/types/proto/meta.proto
Comment (lichuang) [reply common/meta/types/proto/meta.proto:141]: emm, etcd watch API has binary c/s stream implementation, it can save some resource if the client create/cancel/ and then create another watch request again...

File: metasrv/src/watcher/watcher_manager.rs
Comment (ariesdevil) [inline metasrv/src/watcher/watcher_manager.rs:305]: We should report `create` and `cancel` metrics here.

File: common/meta/types/proto/meta.proto
Comment (drmingdrmer) [reply common/meta/types/proto/meta.proto:141]: hmm... I doubt that the resource reduction is measurable in this case...
But your code your decision:DD

File: common/meta/raft-store/src/state_machine/sm.rs
Comment (drmingdrmer) [reply common/meta/raft-store/src/state_machine/sm.rs:592]: I unresolved this comment, because if using `Some(x), None` to represent a `delete` operation, the `UpdateType` can be removed, but it is still there.

I do not quite get it. Are you gonna remove `UpdateType` or still use it?

File: common/meta/types/proto/meta.proto
Comment (drmingdrmer) [reply common/meta/types/proto/meta.proto:77]: Still, it is not changed but it has been marked as **resolved**.
I am confused:(

File: common/meta/types/proto/meta.proto
Comment (drmingdrmer) [inline common/meta/types/proto/meta.proto:89]: It seems it is possible to make it an `Option`:
https://github.com/tokio-rs/prost#field-modifiers

File: common/meta/types/proto/meta.proto
Comment (lichuang) [reply common/meta/types/proto/meta.proto:102]: Watcher may be notified many event at the same time, for example, if watch ["a", "z"], then event ["a"] and ["z"] come in a same internal, these two events can be merge into one response.

File: metasrv/src/watcher/watcher_config.rs
Comment (lichuang) [reply metasrv/src/watcher/watcher_config.rs:35]: Watcher may be notified many event at the same time, for example, if watch ["a", "z"], then event ["a"] and ["z"] come in a same internal, these two events can be merge into one response.

watcher_notify_internal is the batch response time.

File: metasrv/src/watcher/watcher_manager.rs
Comment (lichuang) [reply metasrv/src/watcher/watcher_manager.rs]: maybe metrics can be added next pr?

File: common/meta/raft-store/src/state_machine/sm.rs
Comment (drmingdrmer) [inline common/meta/raft-store/src/state_machine/sm.rs:577]: Won't the result be `None` if it is a successful delete?

File: common/meta/types/proto/meta.proto
Comment (drmingdrmer) [inline common/meta/types/proto/meta.proto:66]: Using positive expression would be better: `UPDATA` is better than `NODELETE`.

File: metasrv/src/watcher/watcher_config.rs
Comment (drmingdrmer) [reply metasrv/src/watcher/watcher_config.rs:35]: But why does it need batching? The client should receive the event ASAP IMHO.

Comment (lichuang) [reply metasrv/src/watcher/watcher_config.rs:35]: well, it's a tradeoff of efficiency and performance, user can config the internal using the `watcher_notify_internal ` config.

if notify ASAP, the server may be too busy in notifying events; if internal too big, may be events came too late.so using `watcher_notify_internal ` to cache the events and batch notify all the events in this internal, and user can config the internal.

Comment (drmingdrmer) [reply metasrv/src/watcher/watcher_config.rs:35]: The batching can be left to be done in the network layer IMHO.
We do not have to worry too much about it.

If application-level batching is really necessary, trying the best to drain a message channel by a limit of the number of messages would be better than by a timer:

https://github.com/datafuselabs/openraft/blob/b50679dcc6eba4e44a7e2106a1919f350676eff4/openraft/src/replication/mod.rs#L483-L509

File: common/meta/types/proto/meta.proto
Comment (drmingdrmer) [inline common/meta/types/proto/meta.proto:85]: We do not need `EventType` anymore.
`(prev.is_some(), current.is_some())` is enough to describe the event type.

File: common/meta/types/proto/meta.proto
Comment (drmingdrmer) [inline common/meta/types/proto/meta.proto:96]: `created` is not necessary anymore if there is no `cancel` response.

File: common/meta/types/proto/meta.proto
Comment (drmingdrmer) [inline common/meta/types/proto/meta.proto:90]: Is this field still in use?

File: metasrv/src/meta_service/raftmeta.rs
Comment (drmingdrmer) [inline metasrv/src/meta_service/raftmeta.rs:827]: Except for the service endpoint layer, these method does not need to know about tonic request.

A `WatchRequest` is quite enough here.

```suggestion
    pub fn create_watcher_stream(&self, request: WatchRequest, tx: WatcherStreamSender) {
        self.watcher.create_watcher_stream(request, tx)
    }
```

File: metasrv/src/watcher/watcher_manager.rs
Comment (drmingdrmer) [inline metasrv/src/watcher/watcher_manager.rs:49]: Isn't it a duplicate of `Event` defined in `meta.proto`?

File: metasrv/src/watcher/watcher_manager.rs
Comment (drmingdrmer) [inline metasrv/src/watcher/watcher_manager.rs:79]: What about merging all `*_rx` into one mpsc channel?

Such an `rx` can be used to receive every kind of event.
Of cause it requires defining the event with an `enum`.
Managing several enum variants may be easier than managing several channels.

And when this `rx` is shutting down, `WatchManager` will be shut down.

It is similar to `close_stream_tx`: a `closing` event can be replaced with shutting down a channel in most cases.

File: common/meta/types/proto/meta.proto
Comment (lichuang) [reply common/meta/types/proto/meta.proto:83]: > Is this field still in use?

i think client still need a id to identify watcher.

File: metasrv/src/watcher/watcher_manager.rs
Comment (drmingdrmer) [reply metasrv/src/watcher/watcher_manager.rs]: Why resolve this comment? 🤔

File: metasrv/src/watcher/watcher_manager.rs
Comment (drmingdrmer) [reply metasrv/src/watcher/watcher_manager.rs]: I do not see why you always click `resolve` before making any update...

File: common/meta/types/proto/meta.proto
Comment (drmingdrmer) [reply common/meta/types/proto/meta.proto:83]: > i think client still need a id to identify watcher.

And what does a client need a watch id for?...

File: metasrv/src/watcher/watcher_stream.rs
Comment (drmingdrmer) [inline metasrv/src/watcher/watcher_stream.rs:71]: A simpler way for closing a stream is to return the error to its caller.

Then the caller could just drop a `WatchStream` to close it.

File: metasrv/src/watcher/watcher_manager.rs
Comment (drmingdrmer) [inline metasrv/src/watcher/watcher_manager.rs:215]: Since create/cancel is removed, the client may not need to be notified anymore?

Comment (lichuang) [reply metasrv/src/watcher/watcher_manager.rs:215]: maybe notified at the first time the watcher been created?

Comment (drmingdrmer) [reply metasrv/src/watcher/watcher_manager.rs]: I don't know.

Usually, when I create a watcher, I'll expect no message to return until something I'm interested in happens.

File: metasrv/src/watcher/watcher_manager.rs
Comment (drmingdrmer) [inline metasrv/src/watcher/watcher_manager.rs:169]: If the `seq` is discarded, when the client receives an event, it may need to retrieve the `seq` with another RPC. Since several meta operations rely on `seq` to guarantee consistency.

File: metasrv/src/watcher/watcher_manager.rs
Comment (drmingdrmer) [inline metasrv/src/watcher/watcher_manager.rs:98]: Just dropping `self.event_tx` is quite enough to notify the core to shut down.

Then at the receiving end:
```rust
            tokio::select! {
                event = self.event_rx.recv() => {
                    if let Some(event) = event {...}
                    else { /* shutdown */ }
```

File: metasrv/src/watcher/watcher_manager.rs
Comment (drmingdrmer) [inline metasrv/src/watcher/watcher_manager.rs:134]: It looks like the filter should be part of the `WatcherStream`, but not part of the `WatchKey`.

File: metasrv/src/watcher/watcher_manager.rs
Comment (drmingdrmer) [inline metasrv/src/watcher/watcher_manager.rs:129]: This may be more suitable in this case:
`let watcher = self.watcher_streams.remove(&id)`.
The removed item will just be returned.

File: metasrv/src/watcher/watcher_manager.rs
Comment (drmingdrmer) [inline metasrv/src/watcher/watcher_manager.rs:124]: It does not need a `select!` anymore. :)

File: metasrv/tests/it/grpc/metasrv_grpc_watch.rs
Comment (lichuang) [reply metasrv/tests/it/grpc/metasrv_grpc_watch.rs:37]: `watcher_client_main` simulate the watcher client, `test_watch_main` simulate the client which update data to metasrv, one for watcher, one for updater.

Comment (drmingdrmer) [reply metasrv/tests/it/grpc/metasrv_grpc_watch.rs:37]: Put what `watcher_client_main` does in the test function body and the simple data-updater in another task.
This will get rid of these several channels for communication.

File: metasrv/tests/it/grpc/metasrv_grpc_watch.rs
Comment (drmingdrmer) [inline metasrv/tests/it/grpc/metasrv_grpc_watch.rs:100]: I mean, put these 3 lines in to a task, then you can just remove `watcher_client_main` along with the channels for communication: just move the body of `watcher_client_main` here.

Comment (lichuang) [reply metasrv/tests/it/grpc/metasrv_grpc_watch.rs:100]: `client.upsert_kv` must wait for watcher has been created, so use the channel to notify the watcher has been created, only after that the `update_kv` has meaning in this case.

Comment (drmingdrmer) [reply metasrv/tests/it/grpc/metasrv_grpc_watch.rs:100]: That's not a problem. You can just create the watcher, then spawn the task, then drain the response stream to check the events.

File: metasrv/tests/it/grpc/metasrv_grpc_watch.rs
Comment (drmingdrmer) [inline metasrv/tests/it/grpc/metasrv_grpc_watch.rs:53]:
```suggestion
    tokio::spawn(upsert_kv_client_main(addr.clone(), updates));
```

File: metasrv/tests/it/grpc/metasrv_grpc_watch.rs
Comment (drmingdrmer) [inline metasrv/tests/it/grpc/metasrv_grpc_watch.rs:73]: Waiting for a single channel does not need a `select!`

File: metasrv/tests/it/grpc/metasrv_grpc_watch.rs
Comment (drmingdrmer) [inline metasrv/tests/it/grpc/metasrv_grpc_watch.rs:63]: This should be enough to guarantee what you expect.
```suggestion
                        assert_eq!(watch_events.remove(0), Some(&event));
```

Comment (lichuang) [reply metasrv/tests/it/grpc/metasrv_grpc_watch.rs]: not a big deal

File: metasrv/tests/it/grpc/metasrv_grpc_watch.rs
Comment (lichuang) [reply metasrv/tests/it/grpc/metasrv_grpc_watch.rs:53]: not a big deal

**Review by ariesdevil — APPROVED** [review]
LGTM, and after this PR merged we need a simple doc that let the cloud team know how to use this feature.

**Comment by drmingdrmer** [comment]
/lgtm

trump-lgtm

**Comment by BohuTANG** [comment]
The API adds/remove will be a compatible problem with old databend-metasrv?
If so, how the user upgrade the old databend-metasrv to the new? There is already some users using databend in the production now :100: 
cc @lichuang @drmingdrmer

**Comment by drmingdrmer** [comment]
> The API adds/remove will be a compatible problem with old databend-metasrv? If so, how the user upgrade the old databend-metasrv to the new? There is already some users using databend in the production now 💯 cc @lichuang @drmingdrmer

This feature does not affect data format thus the upgrade would be smooth:)

**Pull request merged**
